From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.16 october2005, patch 02-15
Patch-Mainline: 2.6.17-rc4

  Description: kernel: signal handling bug.
  Symptom:     Process crashes on signal delivery.
  Problem:     If a signal handler has been established with the SA_ONSTACK
               option but no alternate stack is provided with sigaltstack(),
               the kernel still tries to install the alternate stack. Also
               when setting an alternate stack with sigalstack() and the
               SS_DISABLE flag, the kernel tries to install the alternate
               stack on signal delivery.
  Solution:    Use the correct condition sas_ss_flags() to check if the
               alternate stack has to be used.
  Problem-ID:  23355
  Archs:       s390-31, s390-64

Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.5/arch/s390/kernel/compat_signal.c	20 Mar 2006 15:56:34 -0000	1.22
+++ linux-2.5/arch/s390/kernel/compat_signal.c	21 Apr 2006 17:55:32 -0000	1.22.2.1
@@ -430,7 +430,7 @@
 
 	/* This is the X/Open sanctioned signal stack switching.  */
 	if (ka->sa.sa_flags & SA_ONSTACK) {
-		if (! on_sig_stack(sp))
+		if (! sas_ss_flags(sp))
 			sp = current->sas_ss_sp + current->sas_ss_size;
 	}
 

